// @soos-io/api-client/dist/api/SOOSUserApiClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const constants_1 = require("../constants");
const SOOSApiClient_1 = tslib_1.__importDefault(require("./SOOSApiClient"));
const enums_1 = require("../enums");
class SOOSUserApiClient {
    baseUri;
    apiKey;
    client;
    constructor(apiKey, baseUri = constants_1.SOOS_CONSTANTS.Urls.API.User) {
        this.apiKey = apiKey;
        this.baseUri = baseUri;
        this.client = SOOSApiClient_1.default.create({
            baseUri: this.baseUri,
            apiKey: this.apiKey,
            apiClientName: "User API",
        });
    }
    async getApplicationStatus(clientHash) {
        try {
            const response = await this.client.get(`clients/${clientHash}/application-status`);
            return response.data;
        }
        catch {
            return {
                clientMessage: null,
                statusMessage: {
                    message: "Please verify your API Key and Client ID. Contact support@soos.io if you continue to receive this error.",
                    severity: enums_1.SeverityEnum.High,
                    isDismissible: false,
                    linkText: "",
                    url: "",
                },
            };
        }
    }
}
exports.default = SOOSUserApiClient;
