// @soos-io/api-client/dist/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttributionStatusEnum = exports.AttributionFileTypeEnum = exports.AttributionFormatEnum = exports.FileMatchTypeEnum = exports.HashEncodingEnum = exports.HashAlgorithmEnum = exports.OnFailure = exports.LogLevel = exports.SeverityEnum = exports.ScanType = exports.ScanStatus = exports.ManifestStatus = exports.PackageManagerType = exports.IntegrationType = exports.IntegrationName = exports.ContributingDeveloperSource = void 0;
var ContributingDeveloperSource;
(function (ContributingDeveloperSource) {
    ContributingDeveloperSource["Unknown"] = "Unknown";
    ContributingDeveloperSource["GitHubWebhook"] = "GitHubWebhook";
    ContributingDeveloperSource["EnvironmentVariable"] = "EnvironmentVariable";
    ContributingDeveloperSource["OperatingSystem"] = "OperatingSystem";
})(ContributingDeveloperSource || (exports.ContributingDeveloperSource = ContributingDeveloperSource = {}));
var IntegrationName;
(function (IntegrationName) {
    IntegrationName["AzureDevOps"] = "AzureDevOps";
    IntegrationName["AWSCodeBuild"] = "AWSCodeBuild";
    IntegrationName["Bamboo"] = "Bamboo";
    IntegrationName["BitBucket"] = "BitBucket";
    IntegrationName["CircleCI"] = "CircleCI";
    IntegrationName["CodeShip"] = "CodeShip";
    IntegrationName["GitHub"] = "GitHub";
    IntegrationName["GitLab"] = "GitLab";
    IntegrationName["Jenkins"] = "Jenkins";
    IntegrationName["SoosCsa"] = "SoosCsa";
    IntegrationName["SoosDast"] = "SoosDast";
    IntegrationName["SoosSast"] = "SoosSast";
    IntegrationName["SoosSca"] = "SoosSca";
    IntegrationName["SoosSbom"] = "SoosSbom";
    IntegrationName["SoosScmAudit"] = "SoosScmAudit";
    IntegrationName["TeamCity"] = "TeamCity";
    IntegrationName["TravisCI"] = "TravisCI";
    IntegrationName["VisualStudio"] = "VisualStudio";
    IntegrationName["VisualStudioCode"] = "VisualStudioCode";
})(IntegrationName || (exports.IntegrationName = IntegrationName = {}));
var IntegrationType;
(function (IntegrationType) {
    IntegrationType["None"] = "None";
    IntegrationType["IDE"] = "IDE";
    IntegrationType["Script"] = "Script";
    IntegrationType["Webhook"] = "Webhook";
    IntegrationType["Plugin"] = "Plugin";
    IntegrationType["AppRepo"] = "AppRepo";
    IntegrationType["AppUpload"] = "AppUpload";
})(IntegrationType || (exports.IntegrationType = IntegrationType = {}));
var PackageManagerType;
(function (PackageManagerType) {
    PackageManagerType["Unknown"] = "Unknown";
    PackageManagerType["Alpine"] = "Alpine";
    PackageManagerType["AmazonLinux"] = "AmazonLinux";
    PackageManagerType["CFamily"] = "CFamily";
    PackageManagerType["Dart"] = "Dart";
    PackageManagerType["Debian"] = "Debian";
    PackageManagerType["Docker"] = "Docker";
    PackageManagerType["Erlang"] = "Erlang";
    PackageManagerType["Fedora"] = "Fedora";
    PackageManagerType["GitHub"] = "GitHub";
    PackageManagerType["Go"] = "Go";
    PackageManagerType["Homebrew"] = "Homebrew";
    PackageManagerType["Java"] = "Java";
    PackageManagerType["NPM"] = "NPM";
    PackageManagerType["NuGet"] = "NuGet";
    PackageManagerType["Php"] = "Php";
    PackageManagerType["Python"] = "Python";
    PackageManagerType["Ruby"] = "Ruby";
    PackageManagerType["Rust"] = "Rust";
    PackageManagerType["Swift"] = "Swift";
    PackageManagerType["Wolfi"] = "Wolfi";
    PackageManagerType["Unity"] = "Unity";
})(PackageManagerType || (exports.PackageManagerType = PackageManagerType = {}));
var ManifestStatus;
(function (ManifestStatus) {
    ManifestStatus["Unknown"] = "Unknown";
    ManifestStatus["Valid"] = "Valid";
    ManifestStatus["OnlyDevDependencies"] = "OnlyDevDependencies";
    ManifestStatus["OnlyLockFiles"] = "OnlyLockFiles";
    ManifestStatus["OnlyNonLockFiles"] = "OnlyNonLockFiles";
    ManifestStatus["NoPackages"] = "NoPackages";
    ManifestStatus["UnknownManifestType"] = "UnknownManifestType";
    ManifestStatus["UnsupportedManifestVersion"] = "UnsupportedManifestVersion";
    ManifestStatus["ParsingError"] = "ParsingError";
    ManifestStatus["Empty"] = "Empty";
    ManifestStatus["Duplicate"] = "Duplicate";
})(ManifestStatus || (exports.ManifestStatus = ManifestStatus = {}));
var ScanStatus;
(function (ScanStatus) {
    ScanStatus["Unknown"] = "Unknown";
    ScanStatus["Queued"] = "Queued";
    ScanStatus["NoFiles"] = "NoFiles";
    ScanStatus["Running"] = "Running";
    ScanStatus["LocatingIssues"] = "LocatingIssues";
    ScanStatus["Finished"] = "Finished";
    ScanStatus["FailedWithIssues"] = "FailedWithIssues";
    ScanStatus["Incomplete"] = "Incomplete";
    ScanStatus["Error"] = "Error";
})(ScanStatus || (exports.ScanStatus = ScanStatus = {}));
var ScanType;
(function (ScanType) {
    ScanType["CSA"] = "Csa";
    ScanType["DAST"] = "Dast";
    ScanType["SAST"] = "Sast";
    ScanType["SBOM"] = "Sbom";
    ScanType["SCA"] = "Sca";
})(ScanType || (exports.ScanType = ScanType = {}));
var SeverityEnum;
(function (SeverityEnum) {
    SeverityEnum["Unknown"] = "Unknown";
    SeverityEnum["Info"] = "Info";
    SeverityEnum["Low"] = "Low";
    SeverityEnum["Medium"] = "Medium";
    SeverityEnum["High"] = "High";
    SeverityEnum["Critical"] = "Critical";
})(SeverityEnum || (exports.SeverityEnum = SeverityEnum = {}));
var LogLevel;
(function (LogLevel) {
    LogLevel["DEBUG"] = "DEBUG";
    LogLevel["INFO"] = "INFO";
    LogLevel["WARN"] = "WARN";
    LogLevel["FAIL"] = "FAIL";
    LogLevel["ERROR"] = "ERROR";
})(LogLevel || (exports.LogLevel = LogLevel = {}));
var OnFailure;
(function (OnFailure) {
    OnFailure["Continue"] = "continue_on_failure";
    OnFailure["Fail"] = "fail_the_build";
})(OnFailure || (exports.OnFailure = OnFailure = {}));
var HashAlgorithmEnum;
(function (HashAlgorithmEnum) {
    HashAlgorithmEnum["Unknown"] = "Unknown";
    HashAlgorithmEnum["Md5"] = "Md5";
    HashAlgorithmEnum["Sha1"] = "Sha1";
    HashAlgorithmEnum["Sha256"] = "Sha256";
    HashAlgorithmEnum["Sha512"] = "Sha512";
})(HashAlgorithmEnum || (exports.HashAlgorithmEnum = HashAlgorithmEnum = {}));
var HashEncodingEnum;
(function (HashEncodingEnum) {
    HashEncodingEnum["Utf8"] = "Utf8";
    HashEncodingEnum["Base64"] = "Base64";
    HashEncodingEnum["Binary"] = "Binary";
    HashEncodingEnum["Hex"] = "Hex";
})(HashEncodingEnum || (exports.HashEncodingEnum = HashEncodingEnum = {}));
var FileMatchTypeEnum;
(function (FileMatchTypeEnum) {
    FileMatchTypeEnum["Manifest"] = "Manifest";
    FileMatchTypeEnum["FileHash"] = "FileHash";
    FileMatchTypeEnum["ManifestAndFileHash"] = "ManifestAndFileHash";
})(FileMatchTypeEnum || (exports.FileMatchTypeEnum = FileMatchTypeEnum = {}));
var AttributionFormatEnum;
(function (AttributionFormatEnum) {
    AttributionFormatEnum["Unknown"] = "Unknown";
    AttributionFormatEnum["CsafVex"] = "CsafVex";
    AttributionFormatEnum["CycloneDx"] = "CycloneDx";
    AttributionFormatEnum["Sarif"] = "Sarif";
    AttributionFormatEnum["Spdx"] = "Spdx";
    AttributionFormatEnum["SoosIssues"] = "SoosIssues";
    AttributionFormatEnum["SoosLicenses"] = "SoosLicenses";
    AttributionFormatEnum["SoosPackages"] = "SoosPackages";
    AttributionFormatEnum["SoosVulnerabilities"] = "SoosVulnerabilities";
})(AttributionFormatEnum || (exports.AttributionFormatEnum = AttributionFormatEnum = {}));
var AttributionFileTypeEnum;
(function (AttributionFileTypeEnum) {
    AttributionFileTypeEnum["Unknown"] = "Unknown";
    AttributionFileTypeEnum["Csv"] = "Csv";
    AttributionFileTypeEnum["Html"] = "Html";
    AttributionFileTypeEnum["Json"] = "Json";
    AttributionFileTypeEnum["Text"] = "Text";
    AttributionFileTypeEnum["Xml"] = "Xml";
})(AttributionFileTypeEnum || (exports.AttributionFileTypeEnum = AttributionFileTypeEnum = {}));
var AttributionStatusEnum;
(function (AttributionStatusEnum) {
    AttributionStatusEnum["Unknown"] = "Unknown";
    AttributionStatusEnum["Requested"] = "Requested";
    AttributionStatusEnum["InProgress"] = "InProgress";
    AttributionStatusEnum["Completed"] = "Completed";
    AttributionStatusEnum["CompletedWithProblems"] = "CompletedWithProblems";
    AttributionStatusEnum["Failed"] = "Failed";
})(AttributionStatusEnum || (exports.AttributionStatusEnum = AttributionStatusEnum = {}));
